Structured the "Experimental features" article for completeness & ease of maintenance (#1570)
MicrosoftDocs/edge-developer: microsoft-edge/devtools-guide-chromium/experimental-features/index.md

title: Experimental features
author: MSEdgeTeam
ms.author: msedgedevrel
ms.date: 11/17/2021
ms.topic: article
ms.prod: microsoft-edge
keywords: microsoft edge, web development, f12 tools, devtools, experiment
no-loc: ["Enable webhint", "Enable Network Console", "Source Order Viewer", "Enable Composited Layers in 3D View", "Enable new Font Editor tool within the Styles pane", "Enable new CSS Flexbox debugging features", "Enable + button tab menus to open more tools", "Enable Welcome tab", "3D View", "Turn on support to move tabs between panels", "Match keyboard shortcuts in the DevTools to Microsoft Visual Studio Code", "Edit keyboard shortcuts for any action in the DevTools", "Turn on new CSS grid debugging features", "Emulation: Support dual screen mode"]
---

> # Experimental features

Reviewer: The article should mention, and maybe link to, **Settings** > **Preferences** > the **Restore defaults and refresh** button.

Author: Added

Author: done

Microsoft Edge DevTools provide access to experimental features that are still in development. This article describes only experimental features found in Microsoft Edge. You can try out these features and [provide feedback](../contact.md) before each feature is released.

All [channels of Microsoft Edge](/deployedge/microsoft-edge-channels) have experimental features. You can get the latest experimental features by using the [Microsoft Edge Canary channel](https://www.microsoftedgeinsider.com/welcome?channel=canary). To view the full list available in your version of Microsoft Edge, see the **Settings** > **Experiments** page in DevTools.

<!-- ====================================================================== -->
## Experimental features in the current release

> 

Reviewer: "the current release" is ambiguous.  This article currently lists the experiments in version 98 of Microsoft Edge. 

Might be hard to say "Canary"; that makes a promise about the update schedule for this page.  It's more "timeless" to state the version #.  Would be nice to mention Canary, without promising to keep up to date every month.

Author: Microsoft Edge versions 97 and 98

> The following table lists the default status of each experimental feature. The feature name links to more information.

Reviewer: ```suggestion
The following table lists the default state of each experimental feature. The feature name links to more information.
```
'status' implies "How well is this experiment going?"
'state' or 'setting' would imply checkbox either selected or cleared.  
Maybe change to: 
The default setting for each checkbox in **DevTools** > **Settings** > **Experiments** is shown below.
Note that the UI terminology is "selected" or "cleared", suggesting that the values should read "Selected" and "Cleared", not "On" and "Off" (tho MWSG might say "turn on" in other places) - https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/c/check-checkbox-check-mark - "Use _select_ and _clear_ with checkboxes, not _turn on_ and _turn off_, mark and unmark, check and uncheck, or select and deselect."

Author: deleted

> 

Reviewer: Maybe eliminate the table and place the default state checkbox info in each h2 section below.  The table is manually duplicating the "In this article" navigation feature, difficult to maintain.

Author: fixed

Author: deleted table

> | Default status | Experimental feature |

Reviewer: ```suggestion
| Default state | Experimental feature |
```

| :---: | :--- |

> | Off | [Allow extensions to load custom stylesheets](#allow-extensions-to-load-custom-stylesheets) |

Reviewer: Note that the UI terminology is "selected" or "cleared", suggesting that the values should read "Selected" and "Cleared", not "On" and "Off" (tho MWSG might say "turn on" in other places) - https://docs.microsoft.com/en-us/style-guide/a-z-word-list-term-collections/c/check-checkbox-check-mark - "Use _select_ and _clear_ with checkboxes, not _turn on_ and _turn off_, mark and unmark, check and uncheck, or select and deselect."

Author: deleted table ... but might need to be revisited for the file in general.

Author: table deleted

| Off | [Capture node creation stacks](#capture-node-creation-stacks) |
| Off | [Automatically pretty print in the Sources Panel](#automatically-pretty-print-in-the-sources-panel) |
| Off | [Protocol Monitor](#protocol-monitor) |
| Off | [Show CSP Violations view](#show-csp-violations-view) |
| Off | [Record coverage while performance tracing](#record-coverage-while-performance-tracing) |
| Off | [Show option to take heap snapshot where globals are not treated as root](#show-option-to-take-heap-snapshot-where-globals-are-not-treated-as-root) |
| **On** | [Source order viewer](../accessibility/test-tab-key-source-order-viewer) |
| **On** | [Enable back-forward cache debugging support](#enable-back-forward-cache-debugging-support) |
| Off | [Timeline: event initiators](#timeline-event-initiators) |
| Off | [Timeline: WebGL-based flamechart](#timeline-webgl-based-flamechart) |

> | Off | [Webassembly Debugging: Enable DWARF support](#webassembly-debugging-enable-dwarf-support) |

Reviewer: ```suggestion
| Off | [WebAssembly Debugging: Enable DWARF support](#webassembly-debugging-enable-dwarf-support) |
```

1.  Click the **Settings** (gear) icon to open the [Settings](../customize/index.md#settings) pane of DevTools.

1.  On the left side of the **Settings** pane, click the **Experiments** section.

    :::image type="content" source="../media/experiments-devtools.msft.png" alt-text="The Experiments page in Settings" lightbox="../media/experiments-devtools.msft.png":::

1.  On the **Experiments** page, scroll through the list of all available experimental features. Click an empty checkbox to select each feature that you want to test. Some experiments are turned on by default.

1.  Click a selected checkbox to turn off any feature you do not want to test.

1.  Click the **X** in the upper right to close **Settings**.

1.  Click the **Reload DevTools** button.

> [!NOTE]
> Experimental features are constantly being updated and might cause performance issues. To turn off an experimental feature, open the **Experiments** page and clear the checkbox of the experimental feature that you want to turn off.

<!-- ordering for the h2 sections below: same order as in the Experiments page.  if change to a different ordering scheme, note that scheme here. -->

> 

Reviewer: \#\# Providing feedback about experimental features

Example text from what's new 96:
```
This UI is still in development and may change in future versions of Microsoft Edge.  We're eager to hear your feedback about this new DevTools UI.  Send us your feedback by tweeting [@EdgeDevTools](https://twitter.com/edgedevtools).  Or, with the **Focus Mode** experiment turned on, at the bottom of the **Activity Bar**, select **Help** (![the Help icon in the Activity Bar in Focus Mode](../../../media/help-icon-of-focus-mode.png)) > **Feedback**, to show the **Send feedback** window.
```

Author: Added

> 

Reviewer: \#\# Filtering the experiments

![image](https://user-images.githubusercontent.com/45407486/142060936-b3f42c0b-bc8c-4b86-9ae0-e8cbf6a33833.png)

Author: Added without image.

Author: done, without image

> 

Reviewer: \#\# List of experiments
The following experiments could be unstable or unreliable and may require you to restart DevTools.

ie mirror the UI strings here, that group the checkboxes into 2 groups.

h2 strategy: flattened; since the "In this article" feature shows h2's only, add the two "warning" strings (or equivalent headers) as h2 headings, but keep their children's headings as h2's as well.

Author: Added note style "warning"

<!-- ====================================================================== -->
## Allow extensions to load custom stylesheets

> Placeholder.

Reviewer: Suggested strategy for this PR: in each h2 section, at minimum, have a "See" link, to a reasonable article/section.

Author: Made the suggested change and open to other add-ons links if anyone suggests them.

Placeholder.

<!-- ====================================================================== -->
## Automatically pretty print in the Sources Panel

Placeholder.

<!-- ====================================================================== -->
## Protocol Monitor

Placeholder.

<!-- ====================================================================== -->
## Show CSP Violations view

Placeholder.

<!-- ====================================================================== -->
## Record coverage while performance tracing

Placeholder.

<!-- ====================================================================== -->
## Show option to take heap snapshot where globals are not treated as root

Placeholder.

<!-- ====================================================================== -->
## Enable back forward cache debugging support

Placeholder.

<!-- ====================================================================== -->
## Timeline: event initiators

Placeholder.

<!-- ====================================================================== -->
## Timeline: WebGL based flamechart

Placeholder.

<!-- ====================================================================== -->

> ## Webassembly Debugging: Enable DWARF support

Reviewer: ```suggestion
## WebAssembly Debugging: Enable DWARF support
```

> 

Reviewer: Several headings missing here:
dual screen
APCA
full tree view
font editor

Author: Yes, I was on Beta, not Canary. Will fix.

Placeholder.

<!-- ====================================================================== -->
## Enable automatic contrast issue reporting via the Issues panel

Placeholder.

<!-- ====================================================================== -->
## Enable experimental cookie features

Placeholder.

<!-- ====================================================================== -->
## Enable experimental hide issues menu

Placeholder.

<!-- ====================================================================== -->
## Allow grouping and hiding of issues by IssueKind

Placeholder.

<!-- ====================================================================== -->
## Enable Reporting API panel in the Application panel

Placeholder.

<!-- ====================================================================== -->

> ## Enable CSS \<length\> authoring tool in the Styles pane

Reviewer: @hxlnt - In another PR, you commented "I don't think we picked this up from upstream."  Not sure if there's suggested action here.  The goo.gle link is in the checkbox label in 96 but not later.  We have determined that this page needs to cover the latest Canary checkboxes, each release.  Therefore, the h2 heading shouldn't mention the goo.gle link (because that's been removed from the checkbox label in the latest releases).

Undecided whether the body of this h2 section should show that link, or comment it out (keeping it for authoring), or omit it.

Reviewer 2: In Edge 96, we removed this capability but the (nonfunctional) checkbox remains. This will be removed in subsequent versions of Edge. I think we can just remove this entry entirely.

Author: removed (commented out) in the file.

> ## Enable CSS \<Length\> authoring tool in the styles pane

Reviewer: ```suggestion
## Enable CSS \<length\> authoring tool in the styles pane
```

1.  On the **Experiments** page, scroll through the list of all available experimental features. Click an empty checkbox to select each feature that you want to test. Some experiments are turned on by default.

1.  Click a selected checkbox to turn off any feature you do not want to test.

1.  Click the **X** in the upper right to close **Settings**.

1.  Click the **Reload DevTools** button.

> [!NOTE]
> Experimental features are constantly being updated and might cause performance issues. To turn off an experimental feature, open the **Experiments** page and clear the checkbox of the experimental feature that you want to turn off.

<!-- ordering for the h2 sections below: same order as in the Experiments page.  if change to a different ordering scheme, note that scheme here. -->

<!-- ====================================================================== -->
## Allow extensions to load custom stylesheets

Placeholder.

<!-- ====================================================================== -->
## Capture node creation stacks

> Placeholder.

Author: done

<!-- ====================================================================== -->
## Automatically pretty print in the Sources Panel

Placeholder.

<!-- ====================================================================== -->
## Protocol Monitor

> Placeholder.

Author: Wrote out as step-by-step task.

To use the **Source Order Viewer**:
1.  Open the **Elements** tool.
1.  To the right of the **Styles** tab, click the **Accessibility** tab.
1.  Under the **Source Order Viewer** section, select the **Show Source Order** checkbox.
1.  Highlight any HTML element to display an overlay of the order in the webpage source.

:::image type="content" source="../media/experiments-source-order-viewer.msft.png" alt-text="Source Order Viewer in the Accessibility pane" lightbox="../media/experiments-source-order-viewer.msft.png":::

This experiment is available starting with Microsoft Edge version 86 and is turned on by default.

For more information see [Test keyboard support using the Source Order Viewer](../accessibility/test-tab-key-source-order-viewer.md)

<!-- ====================================================================== -->
## Enable back forward cache debugging support

> Placeholder.

Author: Added

---
# Experimental features

Microsoft Edge DevTools provide access to experimental features that are still in development. This article describes only experimental features found in Microsoft Edge.

All [channels of Microsoft Edge](/deployedge/microsoft-edge-channels) have experimental features. You can get the latest experimental features by using the [Microsoft Edge Canary channel](https://www.microsoftedgeinsider.com/welcome?channel=canary). To view the full list available in your version of Microsoft Edge, see the **Settings** > **Experiments** page in DevTools.

<!-- ====================================================================== -->
## Experimental features which are turned on by default

> As of Microsoft Edge 96, the following experimental features are turned on by default. You can use these features right away, without changing any settings. You can turn off these default experimental features, if needed.

Reviewer: Possibly update article to cover 98 and say 98 here.

1.  Click the **Settings** (gear) icon to open the [Settings](../customize/index.md#settings) pane of DevTools.

1.  On the left side of the **Settings** pane, click the **Experiments** section.

    :::image type="content" source="../media/experiments-devtools.msft.png" alt-text="The Experiments page in Settings" lightbox="../media/experiments-devtools.msft.png":::

1.  On the **Experiments** page, scroll through the list of all available experimental features. Click an empty checkbox to select each feature that you want to test. Some experiments are turned on by default.

1.  Click a selected checkbox to turn off any feature you do not want to test.

1.  Click the **X** in the upper right to close **Settings**.

1.  Click the **Reload DevTools** button.

> [!NOTE]
> Experimental features are constantly being updated and might cause performance issues. To turn off an experimental feature, open the **Experiments** page and clear the checkbox of the experimental feature that you want to turn off.

<!-- ====================================================================== -->
## Restore defaults and refresh

> You can restore the default Experimental features settings.

Reviewer: ```suggestion
To restore the default settings for which experimental features are turned on:
```

> 1.  [Open DevTools](../open/index.md). Click the **Settings and more** button, which is an ellipsis (three dots). Then hover over **More tools**, and click **Developer tools**. Or press **Control + Shift + I**.

Reviewer: ```suggestion
1.  [Open DevTools](../open/index.md).  To do this, in Microsoft Edge, select **Settings and more** (...) > **More tools** > **Developer tools**.  Or press `Ctrl`+`Shift`+`I` (on Windows/Linux) or `Command`+`Option`+`I` (on macOS).
```

Author: yes

 
1.  Click the **Settings** (gear) icon to open the [Settings](../customize/index.md#settings) pane of DevTools.

1.  Click **Preferences** > **Restore defaults and refresh**.

> ### Make sure default settings are restored
> 
> 1.  Open  **Settings** > **Experiments**.
> 
> 1.  Toggle a checkbox in **Experiments**, then close **Settings**.
> 
> 1.  View **Experiments** again and make sure your change persisted.
> 
> 1.  Close **Settings**.
> 
> 1.  Click **Settings** > **Preferences** > **Restore defaults and refresh**.
> 
> 1.  Open **Settings** > **Experiments** again to make sure see the experiment's checkbox reverted to the default state (selected or cleared).

Reviewer: ```suggestion
```
This content is useful for writers but possibly adds more confusion than help (pls assess), when skimming & reading the page.

## Open source files in Visual Studio Code
<!-- in Experiments as of 96, at end of main list -->

The **Open source files in Visual Studio Code** experiment replaces the code editor of the Sources tool with Visual Studio Code, for editing local files. When you turn on this experiment, Developer Tools detects when you edit a local file, and prompts you to select a folder to use as your Workspace.

When you select a folder to use as your Workspace, selecting any link to a file in DevTools opens the file in Visual Studio Code. This action previously opened the file in the code editor of the Sources tool in DevTools.

:::image type="content" source="../media/experiment-sources-in-code-editor-open.msft.png" alt-text="Selecting a file link in the Styles tool opens the file in Visual Studio Code" lightbox="../media/experiment-sources-in-code-editor-open.msft.png":::

Any edits that you make in DevTools now change the file on the hard drive and sync live with Visual Studio Code. You can read about setting up your workspace in [Opening source files in Visual Studio Code](../sources/opening-sources-in-vscode.md).

This experiment is available starting with Microsoft Edge version 96.

> 

Reviewer: Currently (97 & 98) there's a bug, a dup checkbox that starts with "Microsoft Edge ": might want to include here for tracking/completeness: 
Microsoft Edge automatically pretty print in the Sources Panel

Author: I am leaving it off ... if it is here in the next version I will worry about it then.

> <!-- ====================================================================== -->

Reviewer: \#\# Particularly unstable experiments
These experiments are particularly unstable.  Enable at your own risk.

Author: Added

Author: done

## Ignore list for JavaScript frames on Timeline

Placeholder.

<!-- ====================================================================== -->
## Input events on Timeline overview

Placeholder.

<!-- ====================================================================== -->
## Live heap profile

Placeholder.

<!-- ====================================================================== -->
## Sampling heap profiler timeline

Placeholder.

> 

Reviewer: ## Enable keyboard shortcut editor

Author: done

This experiment is available starting with Microsoft Edge version 89.

For more information, see [Edit CSS font styles and settings in the Styles pane in DevTools](../inspect-styles/edit-fonts.md).

<!-- ====================================================================== -->
## Enable automatic contrast issue reporting via the Issues Panel
<!-- present in 96, 98 -->

Needs content.

<!-- ====================================================================== -->
## Enable experimental cookie features
<!-- present in 96, 98 -->

Needs content.

<!-- ====================================================================== -->
## Enable experimental hide issues menu
<!-- present in 96, 98 -->

Needs content.

<!-- ====================================================================== -->
## Allow grouping and hiding of issues by IssueKind
<!-- present in 96, 98 -->

Needs content.

<!-- ====================================================================== -->
## Enable Reporting API panel in the Application panel
<!-- present in 96, 98 -->

> Needs content.

Reviewer 3: Proposal:

"The Reporting API can be used to catch certain errors (like security violations or deprecated API calls) that occur when users visit your site and send them to a server endpoint. Enable this experiment to add the **Reporting API** section in the **Application** panel, which lists all of the reports sent to the endpoint."

Author: added

*   Progressive Web Apps (PWAs)
*   Other common web development issues

The [webhint](https://webhint.io) experiment displays the webhint feedback in the [Issues](../issues/index.md) panel.  Select an issue to display documentation about the solution and a list of the affected resources on your website.  Select a resource link to open the relevant **Network**, **Sources**, or **Elements** pane in DevTools.

:::image type="content" source="../media/experiments-webhint.msft.png" alt-text="webhint feedback in the Issues panel" lightbox="../media/experiments-webhint.msft.png":::

This experiment is available starting with Microsoft Edge version 85 and is turned on by default.

<!-- ====================================================================== -->

## Show issues in Elements
<!-- present in 96, 98 -->

> Needs content.

Reviewer 3: This was documented in What's New for 91 here: https://docs.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/whats-new/2021/04/devtools#wavy-underlines-highlight-code-issues-and-improvements-in-elements-tool
So we could simply link to it.

Author: added short description and link

The **Open source files in Visual Studio Code** experiment replaces the code editor of the Sources tool with Visual Studio Code, for editing local files. When you turn on this experiment, Developer Tools detects when you edit a local file, and prompts you to select a folder to use as your Workspace.

When you select a folder to use as your Workspace, selecting any link to a file in DevTools opens the file in Visual Studio Code. This action previously opened the file in the code editor of the Sources tool in DevTools.

:::image type="content" source="../media/experiment-sources-in-code-editor-open.msft.png" alt-text="Selecting a file link in the Styles tool opens the file in Visual Studio Code" lightbox="../media/experiment-sources-in-code-editor-open.msft.png":::

Any edits that you make in DevTools now change the file on the hard drive and sync live with Visual Studio Code. You can read about setting up your workspace in [Opening source files in Visual Studio Code](../sources/opening-sources-in-vscode.md).

This experiment is available starting with Microsoft Edge version 96.

<!-- ====================================================================== -->

> >> [!WARNING]

Reviewer 3: We will need to remove this I think. There's no content displayed after it, so the warning is not useful anymore.

Author: agreed

<!-- ====================================================================== -->
## DevTools Tooltips
<!-- present in 96, 98 -->

> Needs content.

Reviewer 3: This was also documented in What's New for 91 here: https://docs.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/whats-new/2021/04/devtools#learn-about-devtools-with-informative-tooltips

Author: added

<!-- ====================================================================== -->
## Detached Elements
<!-- present in 96, 98 -->

Memory leaks in web applications are a difficult to locate and repair.

The new Detached Elements panel in Microsoft Edge DevTools helps investigate and resolve a common source of memory leaks.

Memory leaks occur when the JavaScript code of the application retains an increasing number of objects in memory. The JavaScript code should release these objects so that the memory space can be reused by the browser.

<!-- ???more needed, probably with an image, Detached Elements tutorial is in progress??? -->

<!-- ====================================================================== -->
## VS Code themes for the DevTools
<!-- present in 96, 98 -->

> Needs content.

Reviewer 3: Documented in What's New for 93: https://docs.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/whats-new/2021/07/devtools#apply-themes-from-visual-studio-code-to-devtools
And here too: https://docs.microsoft.com/en-us/microsoft-edge/devtools-guide-chromium/customize/theme

I suggest linking to the latter.

Author: added the latter
